src/main/java/io/github/fablabsmc/fablabs/api/fluidvolume/v1/math/Fraction.java

```diff
@@ -1,17 +1,20 @@
-package io.github.fablabsmc.fablabs.api.fluidvolume.v1;
+package io.github.fablabsmc.fablabs.api.fluidvolume.v1.math;
 
-import static com.google.common.base.Preconditions.checkArgument;
-import static com.google.common.math.IntMath.gcd;
-
-import java.util.Objects;
 import java.util.stream.IntStream;
 
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.math.IntMath;
 import com.mojang.datafixers.Dynamic;
 import com.mojang.datafixers.types.DynamicOps;
 
+import net.minecraft.nbt.CompoundTag;
 import net.minecraft.util.DynamicSerializable;
 
+/**
+ * a number who's value is represented as 2 integers, it's numerator and denominator
+ * this class is immutable and self-simplifies on instantiation.
```

Comment:
we may want to split off whole numbers and reduce improper fractions to mixed numbers because I could see this having issues where the numerator gets too big because of some wacky denominator and it hits Integer.MAX

For example, the number of 124,214 buckets isn't *too* insane or hard to deal with, but then if somehow you got a denominator of 21,424, then your numerator is 2,661,160,736‬

Reply:
plus, if ever displayed as a fraction, it will need to be converted to a mixed number anyway.

Reply:
it self simplifies, so the denominator would reduce to 1, unless you mean like int max - 1/2, but that's 3 ints and 3 times the math

Reply:
alternatively we could just slap a long on it and call it a day, also you may want to look at the toString method, which does what you want

Reply:
it self simplified until you have a numerator of 2,661,160,736‬ + 1

Why not store wholes the whole time instead of calculating each time you need it? Would allow for larger numbers and should save some unnecessary calculations

Reply:
I thought we already delegated this arguments to the next PR lol

Reply:
Yeah, this is gonna be for the next PR.